src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -122,47 +112,78 @@ public class EpochManager : IEpochManager {
 		public EpochRecord[] GetLastEpochs(int maxCount) {
 			lock (_locker) {
 				var res = new List<EpochRecord>();
-				for (int epochNum = _lastEpochNumber, n = maxCount; epochNum >= 0 && n > 0; --epochNum, --n) {
-					EpochRecord epoch;
-					if (!_epochs.TryGetValue(epochNum, out epoch))
-						break;
-					res.Add(epoch);
+				var node = _epochs.Last;
+				while (node != null & res.Count < maxCount) {
+					res.Add(node.Value);
+					node = node.Previous;
 				}
-
 				return res.ToArray();
 			}
 		}
 
-		public EpochRecord GetEpoch(int epochNumber, bool throwIfNotFound) {
+		public EpochRecord GetEpochAfter(int epochNumber, bool throwIfNotFound) {
+			if (epochNumber == LastEpochNumber) {
+				if (!throwIfNotFound)
+					return null;
+				throw new ArgumentOutOfRangeException(
+					nameof(epochNumber),
+					$"EpochNumber no epochs exist after Last Epoch {epochNumber}.");
+			}
+
+			EpochRecord epoch;
 			lock (_locker) {
-				if (epochNumber < _minCachedEpochNumber) {
-					if (!throwIfNotFound)
-						return null;
-					throw new ArgumentOutOfRangeException(
-						"epochNumber",
-						string.Format("EpochNumber requested should not be cached. Requested: {0}, min cached: {1}.",
-							epochNumber,
-							_minCachedEpochNumber));
+				var epochNode = _epochs.Last;
+				while (epochNode != null && epochNode.Value.EpochNumber != epochNumber) {
+					epochNode = epochNode.Previous;
 				}
 
-				EpochRecord epoch;
-				if (!_epochs.TryGetValue(epochNumber, out epoch) && throwIfNotFound)
-					throw new Exception(string.Format("Concurrency failure, epoch #{0} should not be null.",
-						epochNumber));
-				return epoch;
+				epoch = epochNode?.Next?.Value;
```

Comment:
GetEpochAfter() currently returns `null` even if the value of `epoch` here is not null.